### ISSUE TITLE
Extension proposal: VK_EXT_descriptor_set_view

### ISSUE THREAD

**Issue by procedural** [KhronosGroup/Vulkan-Docs#1188]
This extension decouples descriptors from descriptor sets, assuming an implementation stores descriptors in descriptor pools' memory and descriptor sets are just views into descriptor pools' memory, allowing for descriptor set overlaps and cheaper descriptor set creation at runtime.

`VK_EXT_descriptor_set_view` adds:

* A `VkBool32` device feature that enables the behaviour described below.
* Explicit descriptor offsets for descriptor sets passed through `vkAllocateDescriptorSets::pAllocateInfo::pNext` extension struct.
* Some optional hardware-specific descriptor pool device limits, such as, for example, "don't mix buffer/image/etc. and sampler descriptors in the same descriptor pool".

Implementation requirements:

* Descriptors that can live in the same descriptor pool are of the same byte size and byte offset between each other.
* Some types of descriptors can't live in the same descriptor pool for hardware-specific reasons exposed in this extension's device limits.

Currently, Vulkan behaves like this:

Say, we create some empty `VkDescriptorPool0` and `VkDescriptorPool1`:

```
↓   is an internal descriptor pool offset
DS. stands for Descriptor Set
D.  stands for Descriptor

↓                                                  ↓
------------------------------------------------   ------------------------
[               VkDescriptorPool0              ]   [   VkDescriptorPool1  ]
------------------------------------------------   ------------------------
```

We allocate descriptor sets `DS.0` `DS.1` from `VkDescriptorPool0` and descriptor set `DS.2` from `VkDescriptorPool1` with `vkAllocateDescriptorSets`:

```
                                               ↓                          ↓
------------------------------------------------   ------------------------
[               VkDescriptorPool0              ]   [   VkDescriptorPool1  ]
------------------------------------------------   ------------------------
[         DS.0         ][         DS.1         ]   [         DS.2         ]
------------------------------------------------   ------------------------
```

We write descriptors `D.00` `D.01` `D.02` to `DS.0`, descriptors `D.03` `D.04` `D.05` to `DS.1` and descriptors `D.06` `D.07` `D.08` to `DS.2` with `vkUpdateDescriptorSets`:

```
                                               ↓                          ↓
------------------------------------------------   ------------------------
[               VkDescriptorPool0              ]   [   VkDescriptorPool1  ]
------------------------------------------------   ------------------------
[         DS.0         ][         DS.1         ]   [         DS.2         ]
------------------------------------------------   ------------------------
[ D.00 ][ D.01 ][ D.02 ][ D.03 ][ D.04 ][ D.05 ]   [ D.06 ][ D.07 ][ D.08 ]
```

If we reset `VkDescriptorPool0` and `VkDescriptorPool1` with `vkResetDescriptorPool`, descriptor sets and their descriptors will be implicitly freed according to the current Vulkan spec:

```
↓                                                  ↓
------------------------------------------------   ------------------------
[               VkDescriptorPool0              ]   [   VkDescriptorPool1  ]
------------------------------------------------   ------------------------
```

But what if descriptors are of the same byte size, stored in descriptor pool's memory tightly or at the same byte offset and are not affected by descriptor pool resets?

Here's what will happen with `VK_EXT_descriptor_pool_memory` enabled:

We create empty `VkDescriptorPool0` and `VkDescriptorPool1`:

```
------------------------------------------------   ------------------------
[      ][      ][      ][      ][      ][      ]   [      ][      ][      ]
------------------------------------------------   ------------------------
[               VkDescriptorPool0              ]   [   VkDescriptorPool1  ]
------------------------------------------------   ------------------------
```

We "allocate" (effectively, just view into the descriptor pools' memory) descriptor sets `DS.0` (at descriptor offset 0) `DS.1` (at descriptor offset 3) from `VkDescriptorPool0` and descriptor set `DS.2` (at descriptor offset 0) from `VkDescriptorPool1` with `vkAllocateDescriptorSets` and descriptor offsets specified in `vkAllocateDescriptorSets::pAllocateInfo::pNext` structure for each descriptor set:

```
------------------------------------------------   ------------------------
[      ][      ][      ][      ][      ][      ]   [      ][      ][      ]
------------------------------------------------   ------------------------
[               VkDescriptorPool0              ]   [   VkDescriptorPool1  ]
------------------------------------------------   ------------------------
[         DS.0         ][         DS.1         ]   [         DS.2         ]
------------------------------------------------   ------------------------
```

We write descriptors `D.00` `D.01` `D.02` to `DS.0`, descriptors `D.03` `D.04` `D.05` to `DS.1` and descriptors `D.06` `D.07` `D.08` to `DS.2` with `vkUpdateDescriptorSets`:

```
------------------------------------------------   ------------------------
[ D.00 ][ D.01 ][ D.02 ][ D.03 ][ D.04 ][ D.05 ]   [ D.06 ][ D.07 ][ D.08 ]
------------------------------------------------   ------------------------
[               VkDescriptorPool0              ]   [   VkDescriptorPool1  ]
------------------------------------------------   ------------------------
[         DS.0         ][         DS.1         ]   [         DS.2         ]
------------------------------------------------   ------------------------

```

But now, if we reset `VkDescriptorPool0` and `VkDescriptorPool1` with `vkResetDescriptorPool`, nothing will happen, descriptor sets and descriptors will not be implicitly freed:

```
------------------------------------------------   ------------------------
[ D.00 ][ D.01 ][ D.02 ][ D.03 ][ D.04 ][ D.05 ]   [ D.06 ][ D.07 ][ D.08 ]
------------------------------------------------   ------------------------
[               VkDescriptorPool0              ]   [   VkDescriptorPool1  ]
------------------------------------------------   ------------------------
[         DS.0         ][         DS.1         ]   [         DS.2         ]
------------------------------------------------   ------------------------
```

Now we can "allocate" (view into the descriptor pools' memory at different descriptor offsets) other descriptor sets that are compatible with previously written descriptors' types:

```
------------------------------------------------   ------------------------
[ D.00 ][ D.01 ][ D.02 ][ D.03 ][ D.04 ][ D.05 ]   [ D.06 ][ D.07 ][ D.08 ]
------------------------------------------------   ------------------------
[               VkDescriptorPool0              ]   [   VkDescriptorPool1  ]
------------------------------------------------   ------------------------
[         DS.0         ][         DS.1         ]   [         DS.2         ]
------------------------------------------------   ------------------------
        [ DS.3 ]        [     DS.4     ]                   [     DS.5     ]
        --------        ----------------                   ----------------
```

We preserved the original behaviour of descriptor sets introduced in Vulkan 1.0, while added the ability to overlap descriptor sets and made descriptor set creation cheaper since now you don't need to write descriptors for the new descriptor sets all over again knowing that they just view already written descriptors in descriptor pool's memory.

Q: Can descriptors of different type be written to already written descriptors in a descriptor pool memory?
A: Yes, since all descriptors in a descriptor pool memory have the same byte size and byte offset between each other, they can be overwritten with descriptors of different type, as long as the descriptor pool allows descriptors of that type according to this extension's device limits.

Q: What will happen if a bound descriptor set specified one of its descriptors to be of one descriptor type, but the underlying descriptor in a descriptor pool's memory is of a different descriptor type at draw or dispatch time?
A: The behaviour is undefined and should result in a device loss, since the user broke the promise when a descriptor of one type was written into a descriptor pool memory and then was accessed as if it was of other type. Validation layer checks should be easy to implement if all descriptors' types in descriptor pool memory are updated at `vkUpdateDescriptorSets` call time.

**Comment by procedural** [comment]
Added some stats to the Q&A section. Not bad, if one wants to target only half of the PC Vulkan user base and knows that specifying a million of descriptors in each shader doesn't hit slow paths in drivers of each IHV.

**Comment by procedural** [comment]
Updated the Q&A section with regards to Github's meaningless emoji reactions.

**Comment by Tobski** [comment]
Hi @procedural - thanks for your suggestion! You've touched on one corner of a set of issues that the WG has been exploring in this area, and your proposed solution isn't a million miles away from some of what we've discussed internally. It's not currently a high priority for us, but it is something we are continuing to evaluate. We'll take on board your feedback, and let you know when it progresses. Any other information/feedback/use cases you or anyone else has would be appreciated. Thanks!

**Comment by krOoze** [comment]
Aliasing feels like it would necessitate some tracking; things could not be just simply freed. And all descriptors types being required to have a same size feels like potential memory waste.

What is the difference here from just using the `VkDescriptorSet`s "correctly"? E.g. from the diagram above, why make `DS.3` alias `DS.0`? Why not just make a `DS.3` and then make another descriptor set with the rest of the descriptors?

**Comment by procedural** [comment]
@Tobski thank you and everyone at WG!

**Comment by procedural** [comment]
@krOoze 

> Aliasing feels like it would necessitate some tracking

I don't see for what reason other than validation?

> things could not be just simply freed

Nothing is simply destroyed or freed in Vulkan, all the standard synchronization requirements would still apply: for example, descriptor set views and descriptor pools should not be destroyed or freed if they're currently in use by a device.

> Why not just make a DS.3 and then make another descriptor set with the rest of the descriptors?

Without aliasing, that another descriptor set would have to include a copy of `DS.3`, bind two descriptor sets (when with descriptor aliasing it would need only one, considering each bound descriptor set has a sensitive memory cost), or both. Explode the example to more descriptors and sets:
```
--------------------------------------------------------------------------------
[ D.00 ][ D.01 ][ D.02 ][ D.03 ][ .... ][ .... ][ .... ][ D.97 ][ D.98 ][ D.99 ]
--------------------------------------------------------------------------------
[                               VkDescriptorPool0                              ]
--------------------------------------------------------------------------------
[                                     DS.0                                     ]
--------------------------------------------------------------------------------
                        [                     DS.3                     ]
                        ------------------------------------------------
[             DS.6             ]
--------------------------------
                [                DS.7                  ]
                ----------------------------------------
                                        [                 DS.8                 ]
                                        ----------------------------------------
                                               [          DS.9         ]
                                               -------------------------
        [             DS.10            ]
        --------------------------------
```
`Shader 1` needs `DS.0`, `Shader 2` needs `DS.3`, `Shader 3` needs `DS.6` and `DS.8`, `Shader 4` needs `DS.7`, `Shader 5` needs `DS.9` and `DS.10`, `Shader 6` needs `DS.6` and `DS.9`, `Shader 7` needs...

**Comment by krOoze** [comment]
> I don't see for what reason other than validation?

Nah, I think I see what you mean. But the `VkDescriptorPool` name is throwing me off. It has no resemblance to the original `Pool`.

I mean, if everything has to have the same size and mutable type, and the pool has to be preallocated in full, then why even have a pool. Could just as well be `std::array<byte>`, and the `sizeof(Descriptor)` could just be queried.

**Comment by procedural** [comment]
> But the VkDescriptorPool name is throwing me off. It has no resemblance to the original Pool.

Yeah... Also, `vkAllocateDescriptorSets` won't allocate anything anymore.

The reason not a single new procedure was introduced here, and all the old ones are reused, is because full backward compatibility was an explicit goal of mine: such an extension must not change the behavior of all the current Vulkan code. If an explicit offset is not passed to `vkAllocateDescriptorSets::pAllocateInfo::pNext`, then `an internal descriptor pool offset` (shown as `↓` in the first diagrams) is used and `vkResetDescriptorPool` just resets the position of that offset back to the beginning of its pool, effectively reproducing the current behavior 1:1.

**Comment by Tobski** [comment]
Now that we've released VK_EXT_descriptor_heap, I think we can safely close this, as all of this is doable as an application within the bounds of descriptor heap, as it is _just_ memory and you can lay it out however you want now.
